### PR TITLE
Convert navigation link to use hooks and context.

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

Follow-up on [this comment](https://github.com/WordPress/gutenberg/pull/28868#discussion_r574195786).

The Navigation link block edit function was using a messy mix of HOCs and hooks for select/dispatch functionality. This PR cleans up and converts it to use only hooks, and get the inherited styles from the parent Navigation block from context instead of attributes.

## How has this been tested?

Tested in browser by adding a Navigation block, some links and some submenus with more links, and checking all is working correctly.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by talldan — APPROVED** [review]
Looks good, no issues in testing, the only thing I noticed was a minor variable naming thing.

File: packages/block-library/src/navigation-link/edit.js [hunk 1]
```diff
@@ -155,18 +142,75 @@ function NavigationLinkEdit( {
...
+				isImmediateParentOfSelectedBlock: hasSelectedInnerBlock(
+					clientId,
+					false
+				),
+				hasDescendants: !! descendants,
+				selectedBlockHasDescendants: !! getClientIdsOfDescendants( [
+					selectedBlockId,
+				] )?.length,
+				numberOfDescendants: descendants,
+				isBeingDragged: isDraggingBlocks(),
```
[the remark this answers is not in the document]
Comment (tellthemachines) [reply packages/block-library/src/navigation-link/edit.js:191]: I couldn't reuse `isDraggingBlocks` without making linter unhappy, and I'm not sure what the best name might be. Is `areBlocksBeingDragged` too verbose?

[the remark this answers is not in the document]
Comment (talldan) [reply packages/block-library/src/navigation-link/edit.js:191]: It's never nice, but the two options are usually assigning the selectors to a variable first and accessing through that:
```jsx
const selectors = select( blockEditorStore );
// ...
isDraggingBlocks: selectors.isDraggingBlocks(),
```

Or prefixing the function with an underscore:
```jsx
const {
    isDraggingBlocks: _isDraggingBlocks,
} = select( blockEditorStore );

return {
    isDraggingBlocks: _isDraggingBlocks(),
}
```

[the remark this answers is not in the document]
Comment (tellthemachines) [reply packages/block-library/src/navigation-link/edit.js:191]: 😑 we can live with the underscore, I guess.

**Pull request merged**